### PR TITLE
[UNR-5020] Custom configs for generated maps

### PR DESCRIPTION
**Pull request by mironec**
Merging into a feature branch.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMapGeneration.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+	UE_LOG(LogTestMapGeneration, Display, TEXT("Deleting the generated test config folder %s."),
+		   *FSpatialTestSettings::GeneratedOverrideSettingsDirectory);
+	if (FPaths::DirectoryExists(FSpatialTestSettings::GeneratedOverrideSettingsDirectory))
+	{
+		IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+		const bool bSuccess = PlatformFile.DeleteDirectoryRecursively(*FSpatialTestSettings::GeneratedOverrideSettingsDirectory);
+		UE_CLOG(!bSuccess, LogTestMapGeneration, Error, TEXT("Failed to delete the generated test config folder %s."),
+				*FSpatialTestSettings::GeneratedOverrideSettingsDirectory);
+	}
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMapGeneration.cpp:32]: New

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMapGeneration.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	{
+		UGeneratedTestMap* TestMap = NewObject<UGeneratedTestMap>(GetTransientPackage(), TestMapClass);
+		TestMap->AddToRoot(); // Okay, must admit, not completely sure what's going on here, seems like even though the commandlet is
+							  // the outer of the newly generated object, the object still gets GCed when creating a new map, so have to add
+							  // to root here to prevent GC
+		if (TestMap->ShouldGenerateMap())
+		{
+			UE_LOG(LogTestMapGeneration, Display, TEXT("Creating the %s."), *TestMap->GetMapName());
+			TestMap->GenerateMap();
+			TestMap->GenerateCustomConfig();
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMapGeneration.cpp:55]: New

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorModule.cpp [hunk 3]
```diff
@@ -52,24 +49,6 @@ void FSpatialGDKEditorModule::StartupModule()
...
-
-#if ENGINE_MINOR_VERSION < 25
-		if (GetDefault<USpatialGDKEditorSettings>()->bStopPIEOnTestingCompleted && GEditor->EditorWorld != nullptr)
-#else
-		if (GetDefault<USpatialGDKEditorSettings>()->bStopPIEOnTestingCompleted && GEditor->IsPlayingSessionInEditor())
-#endif
-		{
-			GEditor->EndPlayMap();
-		}
-	});
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorModule.cpp:72]: Where is this functionality now? Is there a corresponding TestGym PR?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorModule.cpp [hunk 4]
```diff
@@ -300,22 +279,9 @@ bool FSpatialGDKEditorModule::ForEveryServerWorker(TFunction<void(const FName&,
...
-	{
-		FString SnapshotForMap = ASpatialFunctionalTest::GetTakenSnapshotPath(World);
-
-		if (!SnapshotForMap.IsEmpty())
-		{
-			GetMutableDefault<ULevelEditorPlaySettings>()->SetSnapshotOverride(SnapshotForMap);
-			// Set that we're loading from taken snapshot.
-			ASpatialFunctionalTest::SetLoadedFromTakenSnapshot();
-		}
-	}
```
Comment (Vix2021) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorModule.cpp:318]: Where is this functionality now? Is there a corresponding TestGym PR?

**Review by Vix2021 — COMMENTED** [review]
Nice work, just wondering if there is a corresponding TestGym PR?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorModule.cpp (hunk 3 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorModule.cpp:72]: Very sorry about this, realized I forgot to add two files in git.
The functionality is in `SpatialGDKFunctionalTestsModule.cpp` and `SpatialGDKFunctionalTestsModule.h`.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorModule.cpp (hunk 4 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorModule.cpp:318]: Very sorry about this, realized I forgot to add two files in git.
The functionality is in `SpatialGDKFunctionalTestsModule.cpp` and `SpatialGDKFunctionalTestsModule.h`.

**Review by Vix2021 — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/SpatialAuthorityMap.cpp [hunk 5]
```diff
@@ -11,6 +11,7 @@
 USpatialAuthorityMap::USpatialAuthorityMap()
 	: UGeneratedTestMap(EMapCategory::CI_PREMERGE, TEXT("SpatialAuthorityMap"))
 {
+	SetCustomConfig(TEXT("[/Script/UnrealEd.LevelEditorPlaySettings]\nPlayNumberOfClients=1"));
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/SpatialAuthorityMap.cpp:14]:
```suggestion
	SetCustomConfig(TEXT("[/Script/UnrealEd.LevelEditorPlaySettings]" LINE_TERMINATOR
						  "PlayNumberOfClients=1"));
```

Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/SpatialAuthorityMap.cpp:14]: indentation may be wrong, run formatting on this

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Private/SpatialGDKFunctionalTestsModule.cpp [hunk 6]
```diff
@@ -12,8 +17,50 @@ DEFINE_LOG_CATEGORY(LogSpatialFunctionalTest);
...
+	IAutomationControllerManagerPtr AutomationController = AutomationControllerModule.GetAutomationController();
+	AutomationController->OnTestsComplete().AddLambda([]() {
+		// Make sure to clear the snapshot in case something happened with Tests (or they weren't ran properly).
+		ASpatialFunctionalTest::ClearAllTakenSnapshots();
+
+#if ENGINE_MINOR_VERSION < 25
+		if (GetDefault<USpatialGDKEditorSettings>()->bStopPIEOnTestingCompleted && GEditor->EditorWorld != nullptr)
+#else
+		if (GetDefault<USpatialGDKEditorSettings>()->bStopPIEOnTestingCompleted && GEditor->IsPlayingSessionInEditor())
+#endif
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Private/SpatialGDKFunctionalTestsModule.cpp:34]: 4.24 is gone 🔥 
```suggestion
		if (GetDefault<USpatialGDKEditorSettings>()->bStopPIEOnTestingCompleted && GEditor->IsPlayingSessionInEditor())
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Private/SpatialGDKFunctionalTestsModule.cpp [hunk 7]
```diff
@@ -12,8 +17,50 @@ DEFINE_LOG_CATEGORY(LogSpatialFunctionalTest);
...
+
+	FSpatialGDKEditorModule& GDKEditorModule = FModuleManager::LoadModuleChecked<FSpatialGDKEditorModule>(TEXT("SpatialGDKEditor"));
+	GDKEditorModule.OverrideSettingsForTestingDelegate.AddLambda([](UWorld* World, const FString& MapName) {
+		FSpatialGDKFunctionalTestsModule::OverrideSettingsForTesting(World, MapName);
+	});
+}
 
 void FSpatialGDKFunctionalTestsModule::ShutdownModule() {}
 
+void FSpatialGDKFunctionalTestsModule::OverrideSettingsForTesting(UWorld* World, const FString& MapName)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Private/SpatialGDKFunctionalTestsModule.cpp:48]: Looks like this only manages loading from snapshot, should probably update the name?

Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Private/SpatialGDKFunctionalTestsModule.cpp:48]: Unless I misunderstand how this relates to settings

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialTestSettings.cpp [hunk 8]
```diff
@@ -13,6 +13,16 @@ FSpatialTestSettings::FSpatialTestSettings()
 {
 }
 
+const FString FSpatialTestSettings::OverrideSettingsFileExtension = TEXT(".ini");
+const FString FSpatialTestSettings::OverrideSettingsFileDirectoryName = TEXT("MapSettingsOverrides");
+const FString FSpatialTestSettings::OverrideSettingsFilePrefix = TEXT("TestOverrides");
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialTestSettings.cpp:18]: Could we have a separator here so the file names are more readable?
```suggestion
const FString FSpatialTestSettings::OverrideSettingsFilePrefix = TEXT("TestOverrides_");
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMapGeneration.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+namespace TestMapGeneration
+{
+bool GenerateTestMaps()
+{
+	UE_LOG(LogTestMapGeneration, Display, TEXT("Deleting the generated test map folder %s."), *UGeneratedTestMap::GetGeneratedMapFolder());
+	if (FPaths::DirectoryExists(UGeneratedTestMap::GetGeneratedMapFolder()))
+	{
+		IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+		const bool bSuccess = PlatformFile.DeleteDirectoryRecursively(*UGeneratedTestMap::GetGeneratedMapFolder());
+		UE_CLOG(!bSuccess, LogTestMapGeneration, Error, TEXT("Failed to delete the generated test map folder %s."),
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMapGeneration.cpp:20]: Should probably move `bSuccess` to the outer scope to remember if there were any failed steps and return this at the end.
It should probably still be fine to attempt to generate the maps even if deleting the folder failed, though I'm also happy to return early if that fails.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMapGeneration.cpp]: Addressed now

**Pull request merged**